### ISSUE TITLE
`MapperFeature.ALLOW_COERCION_OF_SCALARS = false` should not coerce to String

### ISSUE THREAD

**Issue by driverpt** [FasterXML/jackson-databind#3172]
If we set `MapperFeature.ALLOW_COERCION_OF_SCALARS = false` Jackson will still try to coerce to String
For consistency, this should not happen.

**Comment by cowtowncoder** [comment]
@driverpt I think this is missing a bit of specific problem: which coercion (specifically, which target type, and also which input JSON value shape) is happening that should not? There definitely are cases that are not being checked properly yet but since all checks need to be handled by actual deserializers (some fortunately use shared base impl functionality) it is necessary to know specific cases -- there is no one central place to handle all coercions.

**Comment by driverpt** [comment]
The example is pretty simple.

If your target type is String but if your JSON field is Numeric, it should throw an Exception, not coerce it. This happens for all Scalar Types except for Strings

**Comment by cowtowncoder** [comment]
Ok. I am a bit concerned here as this seems like kind of a change that while technically correct could also effectively break many existing use cases. Implementing the check itself should not be difficult.
But I agree that it should be possible to block this coercion.

**Comment by driverpt** [comment]
I was thinking of creating a PR by adding the `context.hasMapperFeature(ALLOW_COERCION_OF_SCALARS)` or similar to this line of code https://github.com/FasterXML/jackson-databind/blob/2.13/src/main/java/com/fasterxml/jackson/databind/deser/std/StringDeserializer.java#L67

**Comment by cowtowncoder** [comment]
@driverpt With CoercionConfig system, that would not be enough (since there are multiple levels to consider), but the idea itself is sound. You can have a look at `StdDeserializer` for helper methods that do full resolution, for example `_coerceBooleanFromInt()` method.

**Comment by driverpt** [comment]
@cowtowncoder , i think if we do something like:
```
if (t.isScalarValue() && ctxt.hasMapperFeature(ALLOW_COERCION_OF_SCALARS)) {
...
}
```

We should have +/- the expected behaviour. But this might be a breaking change

**Comment by cowtowncoder** [comment]
Right, but this would then completely ignore more accurate `CoercionConfig` system: global `Feature` flags are sort of being deprecated.
Bigger concern however really is the backwards compatibility.

**Comment by driverpt** [comment]
I'm also concerned that this might break backwards-compatibility. What do you suggest then ?

**Comment by cowtowncoder** [comment]
It would be possible to use `CoercionConfig` detection to reliably prevent it; `StringDeserializer` would need to check that.
There are no great examples of this, but if you check `NumberDeserializers.IntegerDeserializer`, it (and similar primitive/wrapper deserializers) calls helper methods in `StdDeserializer` base class.

Checks would need to be done separately for various input "shapes" (boolean, integer, float?), and configuration would need to be done similarly (that is, would need to add separate block for these shapes).

An overview of CoercionConfigs can be found here:

https://cowtowncoder.medium.com/jackson-2-12-most-wanted-4-5-cbc91c00bcd2

which would hopefully help make sense of my explanation.

**Comment by cowtowncoder** [comment]
Would need reproduction (unit test) for specific cases that fail; closing may be reopened with repro.
